### PR TITLE
feat(vended-tools): add notebook to strands-py

### PR DESCRIPTION
**Pull request by chaynabors**
Ports the notebook vended tool from the TypeScript SDK to Python. Gives an agent a session-scoped scratchpad it can create, list, read, write, and clear. Notebooks are stored on agent state under the notebooks key, matching the persistence model of the TypeScript port, so anything the caller already does to persist agent state (Session Management, custom state stores) picks the notebook up for free.

The tool is a thin wrapper over agent state and matches the TypeScript port's naming and semantics with Python snake_case parameter names. Registration lives alongside the bash and file editor tools, exposed via strands.vended_tools.notebook and make_notebook.

Two intentional adds beyond a pure port. First, session caps: sixty-four notebooks per session, one mebibyte per notebook, eight mebibytes total. The TypeScript port has no caps today, but a session-persistent model-driven store needs an upper bound so a prompt injection cannot exhaust memory. Second, notebook-name validation: the model-controlled name becomes a dict key on agent state that may later be serialized to disk, so path-shaped keys are rejected (empty names, path separators, single- and double-dot forms, embedded NULs, over 128 characters, Windows-reserved device names, and invisible format characters). Names are NFKC-normalized before checking so look-alike Unicode cannot bypass validation.

The docs page is updated to add a Python tab for the notebook section, mark the tool as available in Python, and add a Python entry to sourceLinks. TypeScript is otherwise untouched.

https://github.com/strands-agents/harness-sdk/issues/3240

### CODE REVIEW

File: strands-py/src/strands/vended_tools/notebook/notebook.py [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+    # ``in ("..", ".")`` check, but a downstream persister that strips
+    # invisibles then sees plain ``..``.
+    if any(unicodedata.category(c) == "Cf" for c in candidate):
+        raise ValueError("Notebook name must not contain invisible format characters")
+    if "/" in candidate or "\\" in candidate:
+        raise ValueError("Notebook name must not contain path separators")
+    if candidate in ("..", "."):
+        raise ValueError("Notebook name is not allowed")
+    if _WINDOWS_RESERVED.match(candidate):
+        raise ValueError(f"Notebook name '{candidate}' is a reserved device name")
```
[the remark this answers is not in the document]
Comment (chaynabors) [reply strands-py/src/strands/vended_tools/notebook/notebook.py:191]: Fixed in 16a4e3e2dde03011449daed86177422614fb83b8: strip trailing dots and spaces before the reserved-name check, and reject any name with leading or trailing whitespace (or that is whitespace-only). Added parametric tests for CON with a trailing space, CON with a trailing dot, a leading-space CON, a trailing-space notes, and a whitespace-only name.

File: strands-py/src/strands/vended_tools/notebook/notebook.py [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+
+    Returns:
+        A decorated tool that reads and writes notebooks in ``agent.state``.
+    """
+
+    @tool(name=name, description=description, context="tool_context")
+    def notebook_tool(
+        mode: Literal["create", "list", "read", "write", "clear"],
+        tool_context: ToolContext,
+        name: str | None = None,
```
[the remark this answers is not in the document]
Comment (chaynabors) [reply strands-py/src/strands/vended_tools/notebook/notebook.py:67]: Rejected: keeping name= on the factory matches make_bash and make_file_editor, so the two identifiers already coexist across the vended-tools package by convention. Renaming just here would break parity for readers moving between the tools; renaming everywhere is a separate cleanup, not this port.

File: strands-py/tests/strands/vended_tools/test_notebook.py [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+
+
+class TestCreate:
+    """The ``create`` operation."""
+
+    @pytest.mark.asyncio
+    async def test_creates_empty_notebook_with_default_name(self, alist):
+        state, agent = _fresh_agent()
+        result = await _invoke(agent, {"mode": "create"}, alist)
+        assert result["status"] == "success"
```
[the remark this answers is not in the document]
Comment (chaynabors) [reply strands-py/tests/strands/vended_tools/test_notebook.py]: Fixed in 16a4e3e2dde03011449daed86177422614fb83b8: the two stream()-path create tests now assert the full tool_result shape via a single equality. Left the other stream tests as-is (they check a single specific field on top of state assertions, which reads clearly); happy to fold more if reviewers want.

File: strands-py/src/strands/vended_tools/notebook/types.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    Notebooks are stored in the agent's state under the ``notebooks`` key.
+    Each notebook stores plain text content with newline-separated lines.
+    """
+
+    notebooks: dict[str, str]
+
+
+DEFAULT_NOTEBOOK_DESCRIPTION = (
+    "Manages text notebooks for note-taking and documentation. Supports create, list, read, "
+    "write (replace or insert), and clear operations. Notebooks persist within the agent invocation."
```
[the remark this answers is not in the document]
Comment (chaynabors) [reply strands-py/src/strands/vended_tools/notebook/types.py]: Fixed in 16a4e3e2dde03011449daed86177422614fb83b8: the model-facing description and the tool-function docstring both now say notebooks persist across invocations within a session, and across sessions when the agent has a durable state store. The TypeScript port has the same understated wording today; scoping the string fix to Python for this port and will file a follow-up TS PR to bring the two back into sync.

File: strands-py/src/strands/vended_tools/notebook/README.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+agent('Create a notebook called "ideas" with "# Project Ideas"')
+agent('Add "- Build a web scraper" to the ideas notebook')
+agent('Read the ideas notebook')
+```
+
+State is accessible directly:
+
+```python
+print(agent.state.get("notebooks"))
+# {'ideas': '# Project Ideas\n- Build a web scraper'}
```
[the remark this answers is not in the document]
Comment (chaynabors) [reply strands-py/src/strands/vended_tools/notebook/README.md]: Fixed in fe9d2a685fa338281145192d927e89478e9e1d88: updated the README state example to include the auto-materialized default notebook and added a short note explaining why it appears.

File: site/src/content/docs/user-guide/concepts/tools/vended-tools.mdx [hunk 6]
```diff
@@ -97,22 +98,61 @@ This tool makes HTTP requests to arbitrary URLs without restrictions on destinat
...
+agent('Create a notebook called "tasks" with "# Daily Tasks" and add "- [ ] Review code" to it')
+
+# Read the notebook contents directly off agent state.
+print(agent.state.get("notebooks")["tasks"])
+```
+
+</Tab>
+</Tabs>
+
 📖 [Full API Reference](https://github.com/strands-agents/harness-sdk/blob/main/strands-ts/src/vended-tools/notebook/README.md)
```
[the remark this answers is not in the document]
Comment (chaynabors) [reply site/src/content/docs/user-guide/concepts/tools/vended-tools.mdx]: Fixed in fe9d2a685fa338281145192d927e89478e9e1d88: the notebook Full API Reference line now links both the TypeScript and Python READMEs.

File: strands-py/tests/strands/vended_tools/test_notebook.py [hunk 7]
```diff
@@ -0,0 +1,522 @@
...
+
+class TestList:
+    """The ``list`` operation."""
+
+    @pytest.mark.asyncio
+    async def test_lists_default_when_initialized(self, alist):
+        _, agent = _fresh_agent({"default": ""})
+        result = await _invoke(agent, {"mode": "list"}, alist)
+        assert result["status"] == "success"
+        assert "default: Empty" in result["content"][0]["text"]
```
[the remark this answers is not in the document]
Comment (chaynabors) [reply strands-py/tests/strands/vended_tools/test_notebook.py]: Fixed in fe9d2a685fa338281145192d927e89478e9e1d88: the four stream()-path tests (list, read, write, clear defaults) now assert the full tool_result via a single equality.

File: strands-py/tests/strands/vended_tools/test_notebook.py [hunk 8]
```diff
@@ -0,0 +1,522 @@
...
+
+class TestRead:
+    """The ``read`` operation."""
+
+    @pytest.mark.asyncio
+    async def test_reads_entire_notebook_default(self, alist):
+        _, agent = _fresh_agent({"default": "Line 1\nLine 2\nLine 3\nLine 4\nLine 5"})
+        result = await _invoke(agent, {"mode": "read"}, alist)
+        assert result["status"] == "success"
+        assert result["content"][0]["text"] == "Line 1\nLine 2\nLine 3\nLine 4\nLine 5"
```
[the remark this answers is not in the document]
Comment (chaynabors) [reply strands-py/tests/strands/vended_tools/test_notebook.py]: Fixed in fe9d2a685fa338281145192d927e89478e9e1d88: full-object assertion.

File: strands-py/tests/strands/vended_tools/test_notebook.py [hunk 9]
```diff
@@ -0,0 +1,522 @@
...
+    @pytest.mark.asyncio
+    async def test_replaces_text_default_notebook(self, alist):
+        state, agent = _fresh_agent({"default": "# Todo List\n\n[ ] Task 1\n[ ] Task 2\n[x] Task 3"})
+        result = await _invoke(
+            agent,
+            {"mode": "write", "old_str": "[ ] Task 1", "new_str": "[x] Task 1"},
+            alist,
+        )
+        assert result["status"] == "success"
+        assert result["content"][0]["text"] == "Replaced text in notebook 'default'"
```
[the remark this answers is not in the document]
Comment (chaynabors) [reply strands-py/tests/strands/vended_tools/test_notebook.py]: Fixed in fe9d2a685fa338281145192d927e89478e9e1d88: full-object assertion.

File: strands-py/tests/strands/vended_tools/test_notebook.py [hunk 10]
```diff
@@ -0,0 +1,522 @@
...
+
+class TestClear:
+    """The ``clear`` operation."""
+
+    @pytest.mark.asyncio
+    async def test_clears_default_notebook(self, alist):
+        state, agent = _fresh_agent({"default": "Some content"})
+        result = await _invoke(agent, {"mode": "clear"}, alist)
+        assert result["status"] == "success"
+        assert result["content"][0]["text"] == "Cleared notebook 'default'"
```
[the remark this answers is not in the document]
Comment (chaynabors) [reply strands-py/tests/strands/vended_tools/test_notebook.py]: Fixed in fe9d2a685fa338281145192d927e89478e9e1d88: full-object assertion.

File: strands-py/src/strands/vended_tools/notebook/notebook.py [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+    MAX_TOTAL_SIZE_BYTES,
+)
+
+if TYPE_CHECKING:
+    from ...tools.decorator import DecoratedFunctionTool
+
+_STATE_KEY = "notebooks"
+
+# Modes that alter ``notebooks`` state. Only these persist a write.
+_MUTATING_MODES = frozenset({"create", "write", "clear"})
```
Comment (yonib05) [inline strands-py/src/strands/vended_tools/notebook/notebook.py:36]: TS `set`s state unconditionally and has a passing test that `list` leaves `{default: ''}` behind, while this only persists on create/write/clear and asserts the opposite. I think the Python behavior is better, but right now the two suites encode contradictory contracts and #3240 asks for verified parity. Fix TS in this PR?

File: strands-py/src/strands/vended_tools/notebook/notebook.py [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+    Notebook names are opaque keys into agent state; they must not resemble
+    filesystem paths and must fit within a reasonable size. The state layer is
+    memory-only by default, so path traversal is not a direct risk, but a name
+    that looks like a path is a footgun for consumers who later persist state.
+
+    Args:
+        candidate: The proposed notebook name.
+
+    Returns:
+        The NFKC-normalized notebook name.
```
Comment (yonib05) [inline strands-py/src/strands/vended_tools/notebook/notebook.py:155]: The notebook name is a key inside a JSON blob in agent state and never touches a path. What are the NFKC normalization, `Cf`-category rejection, and Windows-reserved regex defending against? The normalization also makes a pre-existing NFD key unreachable, since lookup normalizes to NFC and misses.

File: strands-py/src/strands/vended_tools/notebook/types.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
+"""Shared types and constants for the notebook tool."""
+
+from typing import TypedDict
+
+
+class NotebookState(TypedDict):
```
Comment (yonib05) [inline strands-py/src/strands/vended_tools/notebook/types.py:6]: `NotebookState` isn't imported anywhere, and it declares `{notebooks: dict[str, str]}` where we actually store a bare `dict[str, str]`. Looks like it's copying a TS type bug, can we drop it rather than make it public in a second language?

File: strands-py/src/strands/vended_tools/notebook/notebook.py [hunk 14]
```diff
@@ -0,0 +1,353 @@
...
+_GROWING_MODES = frozenset({"create", "write"})
+
+# Windows-reserved device names — refuse them regardless of platform so a session
+# state written on Linux does not become un-persistable on Windows. Real Windows
+# strips any extension before comparing against the device-name list, so
+# ``CON.txt`` and ``NUL.log`` are just as broken as ``CON``.
+_WINDOWS_RESERVED = re.compile(r"^(CON|PRN|AUX|NUL|COM[1-9]|LPT[1-9])(\..*)?$", re.IGNORECASE)
+
+
+def make_notebook(
```
Comment (yonib05) [inline strands-py/src/strands/vended_tools/notebook/notebook.py:48]: `make_notebook` only takes name and description, no behavioral knob, and there's no `makeNotebook` in TS. Add the factory when there's something to configure?

**Review by yonib05 — COMMENTED** [review]
This is conflicting on `vended_tools/__init__.py` now that `http_request` and `sleep` have landed. Mind rebasing and re-applying just the two-line change instead of rewriting the module docstring? Makes the diff easier to read too.

---

The session caps and the `old_str` + `insert_line` ambiguity rejection are both improvements, but they're Python-only right now. Port them to TS here so the two don't fork?

---

These tests mirror the TS cases class for class, including the `$&`-literal and partial-match edge cases. That's the pattern I'd want the other twelve to follow.

**Comment by liramon2** [comment]
Closed in favor of #4132 and #4281
